### PR TITLE
feat(ui): Initial AppearanceProvider support

### PR DESCRIPTION
**Pull request by dstaley**
## Description

This PR contains a complete implementation of the `appearance` prop for the new AIO components. Notable changes from the existing `appearance` prop include the dropping of arbitrary CSS (such as pseudo elements) and the removal of component-specific styles like `appearance={{ elements: { signIn: { ... } } }}`.

This PR also contains the following sub-PRs:
- https://github.com/clerk/javascript/pull/4012
- https://github.com/clerk/javascript/pull/4015
- https://github.com/clerk/javascript/pull/3999
- https://github.com/clerk/javascript/pull/3988

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/ui/src/contexts/AppearanceContext.tsx [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    it('merges object values for the same element', () => {
+      const a = { elements: { alert__warning: { background: 'tomato' } } };
+      const b = { elements: { alert__warning: { color: 'red' } } };
+      expect(mergeAppearance(a, b)).toStrictEqual({
+        layout: {},
+        elements: {
+          alert__warning: { color: 'red', background: 'tomato' },
+        },
+      });
+    });
```
Comment (alexcarpenter) [inline packages/ui/src/contexts/AppearanceContext.tsx:339]: should we add a test where to style objects contain a color key? Does the one farthest down the tree take precedence?

File: packages/ui/src/primitives/alert.tsx [hunk 2]
```diff
@@ -1,29 +1,63 @@
...
+  },
+  alert__warning: {
+    className: 'text-warning bg-warning/[0.06] border-warning/[0.12]',
+  },
+  alert__error: {
+    className: 'text-danger bg-danger/[0.06] border-danger/[0.12]',
+  },
+  alertRoot: {},
+  alertIcon: {},
+} satisfies ParsedElementsFragment;
```
[the remark this answers is not in the document]
Comment (dstaley) [reply packages/ui/src/primitives/alert.tsx:37]: The empty objects are a type checking requirement, but after talking with Bryce we decided to move to a runtime validation approach instead. In a future PR, we'll be able to remove these empty objects, and our `useAppearance` hook will automatically create the empty objects without us needing to specify them.

File: packages/ui/src/contexts/AppearanceContext.tsx (hunk 1 above)
Comment (dstaley) [reply packages/ui/src/contexts/AppearanceContext.tsx:339]: good catch! https://github.com/clerk/javascript/pull/3976/commits/1cdd7ffaec21b4499fecc6e2232447ec02f10ca5

File: packages/ui/src/components/sign-in/steps/choose-strategy.tsx [hunk 3]
```diff
@@ -6,38 +6,25 @@ import { Connections } from '~/common/connections';
...
   ============================================ */
 
-function FirstFactorConnections({
-  isGlobalLoading,
-  hasConnection,
-}: {
-  isGlobalLoading: boolean;
-  hasConnection: boolean;
-}) {
-  const { t } = useLocalizations();
```
Comment (BRKalow) [inline packages/ui/src/components/sign-in/steps/choose-strategy.tsx:30]: Are these changes meant to be here? 👀

Comment (BRKalow) [reply packages/ui/src/components/sign-in/steps/choose-strategy.tsx:30]: Ah, I see yes! Alex had a few PRs targeting this branch 👍

File: packages/ui/src/contexts/AppearanceContext.tsx [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+
+    return { value: { parsedAppearance, themelessAppearance, theme: props.theme ?? globalTheme } };
+  }, [props.appearance, props.theme, globalAppearance, globalThemelessAppearance, globalTheme]);
+
+  return <AppearanceContext.Provider value={ctxValue}>{props.children}</AppearanceContext.Provider>;
+};
+
+export { AppearanceProvider, useAppearance };
+
+if (import.meta.vitest) {
```
Comment (BRKalow) [inline packages/ui/src/contexts/AppearanceContext.tsx:272]: love that you can test local functions without having to export them!

File: packages/ui/src/contexts/AppearanceContext.tsx [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+  }
+
+  if (!result) {
+    throw new Error(`Unable to merge ElementsAppearanceConfigs: ${a} and ${b}`);
+  }
+
+  return result;
+}
+
+function mergeAppearance(a: Appearance | null | undefined, b: Appearance | null | undefined): Appearance | null {
```
Comment (BRKalow) [inline packages/ui/src/contexts/AppearanceContext.tsx:153]: Might be helpful to outline the merging heuristic in prose in a JSDoc. 👀

**Review by alexcarpenter — APPROVED** [review]
This is a solid foundation to build from. I am interested in getting this in sooner than later, and continue iterating on it.

**Review by BRKalow — APPROVED** [review]
👏 great start!

**Pull request merged**